test/tests/p5.SoundRecorder.js

```diff
@@ -0,0 +1,142 @@
+        done();
+      });
+    });
+
+    after(function() {
+      inputSoundFile.dispose();
+      p5.prototype.writeFile = originalWriteFile;
+    });
+
+    beforeEach(function() {
+      recorder = new p5.SoundRecorder();
+      inputSoundFile.disconnect();
+      inputSoundFile.stop();
+      writeFileCalled = false;
+    });
+
+    afterEach(function() {
+      recorder.dispose();
+    });
+
+    it('can record input from a microphone', function(done) {
+      var recordingDuration = 100;
+
+      // need to enable master volume to test recording from the microphone
+      p5.prototype.masterVolume(1);
+
+      var mic = new p5.AudioIn();
+      mic.start(function() {
+        var outputSoundFile = new p5.SoundFile();
+        recorder.record(outputSoundFile);
+
+        setTimeout(function() {
+          recorder.stop();
+          expect(outputSoundFile.duration()).to.be.closeTo(recordingDuration / 1000, 0.1);
+
+          var outputChannel = outputSoundFile.buffer.getChannelData(0);
+          expect(outputChannel[0]).to.not.eq(0);
+
+          outputSoundFile.dispose();
+          mic.dispose();
+          p5.prototype.masterVolume(0);
+          done();
+        }, recordingDuration);
+      });
+    });
+
+    it('can record input from a sound file', function(done) {
+      // TODO: why does a sampleIndex of 0 break this test?
+      var sampleIndex = 1000;
+      var recordingDuration = 100;
+      var inputChannel = inputSoundFile.buffer.getChannelData(0);
+      // input SoundFile should contain all 1s
+      expect(inputChannel[sampleIndex]).to.eq(1);
+
+      var outputSoundFile = new p5.SoundFile();
+      inputSoundFile.play();
+      recorder.setInput(inputSoundFile);
+      recorder.record(outputSoundFile);
+
+      setTimeout(function() {
+        recorder.stop();
+        expect(outputSoundFile.duration()).to.be.closeTo(recordingDuration / 1000, 0.1);
+
+        var outputChannel = outputSoundFile.buffer.getChannelData(0);
+        expect(outputChannel[sampleIndex]).to.eq(1);
+
+        outputSoundFile.dispose();
+        done();
+      }, recordingDuration);
+    });
+
+    it('can record the master output of a sketch', function(done) {
+      var recordingDuration = 100;
+      var inputChannel = inputSoundFile.buffer.getChannelData(0);
+      // input SoundFile should contain all 1s
+      expect(inputChannel[0]).to.eq(1);
+
+      // need to enable master volume to test recording from master output
+      p5.prototype.masterVolume(1);
+
+      var outputSoundFile = new p5.SoundFile();
+      inputSoundFile.connect();
+      inputSoundFile.play();
+      recorder.setInput();
+      recorder.record(outputSoundFile);
+
+      setTimeout(function() {
+        recorder.stop();
+        expect(outputSoundFile.duration()).to.be.closeTo(recordingDuration / 1000, 0.1);
+
+        var outputChannel = outputSoundFile.buffer.getChannelData(0);
+        expect(outputChannel[0]).to.not.eq(0);
```

Comment:
maybe we can be more precise here if it's a generated buffer

Reply:
Even with a generated buffer, I'm still having issues testing against precise values here. It seems like playing the input SoundFile through the master soundOut node produces a different output buffer, even with the master volume at 1. I don't really understand why...

Reply:
Ah, we have a pretty aggressive dynamicCompressorNode inside of the Master that is probably clamping the values. Perhaps a value of 0.5 would pass thru just fine?

Reply:
Ah, we have a pretty aggressive dynamicCompressorNode inside of the Master that is probably clamping the values. Perhaps a value of 0.5 would pass thru just fine?

Reply:
Ah, that's probably it.